### PR TITLE
Add needs-unwind to tests that depend on panicking

### PR DESCRIPTION
**Pull request by djkoloski**
These tests were found by running the test suite on fuchsia which compiles with `panic=abort` by default, then picking through the failures manually to locate the tests that require unwinding support.

Most of these tests are already opted-out on platforms that compile with `panic=abort` by default. This just generalizes it a bit more so that fuchsia tests can be run properly. Currently, the `needs-unwind` directive needs to be manually passed to compiletest (e.g. via `--test-args '--target-panic=abort'`). Eventually, I would like `x.py` or compiletest to determine whether the directive should be used automatically based on the target panic settings.

### CODE REVIEW

**Comment by tmandry** [comment]
You should also be able to remove the existing `ignore-(platform)` lines that are there because the platform is built without unwinding support.

**Comment by djkoloski** [comment]
I did that originally but the flag in compiletest needs a command-line argument (i.e. it's not inherently set for specific target configurations). Should we be pulling that setting from a config?

**Comment by tmandry** [comment]
Ah right, the plan was to add a flag to bootstrap that enables `panic=abort` and tells compiletest about it. That should be a separate change. I think this is fine for now.

File: src/test/ui/async-await/async-fn-size-moved-locals.rs [hunk 1]
```diff
@@ -8,6 +8,7 @@
 // See issue #59123 for a full explanation.
 
 // ignore-emscripten (sizes don't match)
+// ignore-fuchsia (sizes don't match exactly)
```
Comment (Mark-Simulacrum) [inline src/test/ui/async-await/async-fn-size-moved-locals.rs:11]: This seems unrelated to the PR's stated purpose, I'd prefer to split it (at least to a separate commit, ideally PR). I'm not sure why emscripten and fuchsia are special here, that seems surprising to me. (Maybe this is a needs-unwind in disguise?)

**Comment by Mark-Simulacrum** [comment]
> I did that originally but the flag in compiletest needs a command-line argument (i.e. it's not inherently set for specific target configurations). Should we be pulling that setting from a config?

I forget what PR added needs-unwind (maybe my bad there), but if it's currently just not being used at all (which seems to be the case), it seems a little misleading to go through annotating tests with this flag, particularly when it "seems" like some of the ignore- flags would otherwise be removable.

I also noted that the current target specs for fuchsia don't have a panic strategy of abort, so it seems like in general unwinding is supported?

I am not strictly opposed to merging this, but I would like a better understanding of the expectation on your end for integrating needs-unwind into actual invocations of compile test from our build system (rough sense of timeline), and maybe some commentary in the PR description for how these tests were found (did you run with a panic=abort flag and just blanket added for where it failed?). My review comment suggests that it's possible some unrelated tests were also adjusted in this PR, which seems best to avoid and is part of the reason I'd like the test selection rationale better documented in the PR.

File: src/test/ui/async-await/async-fn-size-moved-locals.rs (hunk 1 above)
Comment (djkoloski) [reply src/test/ui/async-await/async-fn-size-moved-locals.rs:11]: This was accidentally included, I'll split this into a separate PR.

**Comment by djkoloski** [comment]
> if it's currently just not being used at all (which seems to be the case), it seems a little misleading to go through annotating tests with this flag

There is a flag and it is functional, but it has to be manually passed (e.g. `./x.py test --test-args '--target-panic=abort'`). It should really be pulled from the target specification instead, but I haven't done any work to get that up and running yet. Once the flag is automatically specified, I would be comfortable removing the target-specific `ignore-` directives.

> I also noted that the current target specs for fuchsia don't have a panic strategy of abort, so it seems like in general unwinding is supported?

Although I believe it is now possible to build fuchsia with unwinding, I believe this was not always the case and the standard library is compiled with `panic=abort` by default. I can check with @tmandry on this.

> I would like a better understanding of the expectation on your end for integrating needs-unwind into actual invocations of compile test

I expect that eventually x.py will automatically determine whether the `needs-unwind` directive applies to a target. Getting the Rust test suite running on fuchsia is my primary objective.

> maybe some commentary in the PR description for how these tests were found

Updated.

> My review comment suggests that it's possible some unrelated tests were also adjusted in this PR

Ack, you are correct. Those are fixes for other tests that failed on fuchsia, but are unrelated to this PR. They were added by mistake and I've removed them from this PR.

**Comment by tmandry** [comment]
For context, my plan was to add an explicit flag to x.py for building with panic=abort (ideally with defaults for the platforms that don't support it). This would modify compilation flags (including `-Zpanic-abort-tests` in the case of testing) as well as cause `--target-panic=abort` to be passed to compiletest.

It's true that adding the flag while it's not fully functional could be misleading, but at least for now there wouldn't be any examples where it was there without the right platforms also being ignored explicitly. (Eventually we would take those out.)

**Comment by Mark-Simulacrum** [comment]
I'm okay proceeding here -- it sounds like we're on a good path toward this being integrated, and it's already ~usable directly, when needed.

@djkoloski it looks like an unrelated commit and a couple merges landed on this branch -- can you rebase those away? After that's done I expect to r+
